### PR TITLE
Add lookahead function

### LINKED ISSUE #136: Implement Parsimmon.lookahead
Hi, I'm very fan of this project.
I'm using this in my product [markdown-next](https://github.com/sketchglass/markdown-next).
At that time I had to implement function like notRegexp.
If you do not mind I'd like to implement `Parsimmon.not` method, which inverts argument parser's sucess and failure.

### PR DESCRIPTION
**Pull request by minamorl**
closes #136 
- string and regex support
- provides as P.lookahead and P.prototype.lookahead for convenience

### CODE REVIEW

File: src/parsimmon.js [hunk 1]
```diff
@@ -494,6 +497,30 @@
...
+        var j = i + x.length;
+        var head = input.slice(i, j);
+        if (head === x) {
+          return makeSuccess(i, '');
+        } else {
+          return makeFailure(i, x);
+        }
+      });
+    } else if (x instanceof RegExp) {
+      var regexp = RegExp('^(?:' + x.source + ')', flags(x));
```
Comment (wavebeem) [inline src/parsimmon.js]: See `assertRegexp`, certain regexp flags are not allowed in Parsimmon, so we should use that function here.

File: src/parsimmon.js [hunk 2]
```diff
@@ -494,6 +497,30 @@
...
+      return Parsimmon(function(str, i) {
+        if (regexp.test(str.slice(i))) {
+          return Parsimmon.makeSuccess(i, '');
+        }
+        return Parsimmon.makeFailure(i, '' + regexp);
```
Comment (wavebeem) [inline src/parsimmon.js]: You can remove the `Parsimmon.` on these `makeSuccess` and `makeFailure`

File: src/parsimmon.js [hunk 3]
```diff
@@ -494,6 +497,30 @@
...
+      });
+    }
+    throw new Error('not a string or regexp: '+x);
```
Comment (wavebeem) [inline src/parsimmon.js]: A space around the `+` would be nice 👍

File: test/parsimmon.test.js [hunk 4]
```diff
@@ -120,6 +121,48 @@ suite('parser', function() {
...
+      var answer = parser.parse('abcd');
+      assert.deepEqual(answer.value, ['abc', '', 'd']);
+    });
+    test('raises error if argument is not string or regexp', function() {
+      try {
+        lookahead(new Object());
+      } catch (err) {
+        true;
+      }
+    });
```
Comment (wavebeem) [inline test/parsimmon.test.js]: The function `assert.throws` is helpful for this kind of test:

```javascript
assert.throws(function() { lookahead({}); });
assert.throws(function() { lookahead([]); });
assert.throws(function() { lookahead(12); });
assert.throws(function() { lookahead(true); });
// etc...
```

**Comment by minamorl** [comment]
Thanks for your reviewing.

> Also, I wonder if we really need both P.lookahead and P.prototype.lookahead. I think the second case makes a lot more sense, and it's probably less confusing if there's only one.

I personally think this should be allowed to use `lookahead` as an independent function, for example, I want to use `lookahead` inside `P.seq()` and `P.alt()` for readability and maintainability.
But I respect your design, so if you want to drop support Parsimmon.lookahead, I can change.

**Comment by wavebeem** [comment]
You could always make your own helper like:

```javascript
function lookahead(x) {
  return P.string('').lookahead(x);
}
```

Give me some time to think about it. I can see your point. I might change my mind later.

**Comment by wavebeem** [comment]
I decided to keep both functions. Thanks for your contribution!!

**Pull request merged**